### PR TITLE
refactor(mobile): Use ImmichThumbnail and local thumbnail image provider

### PR DESCRIPTION
**Pull request by martyfuhry**
- Makes a new `ImmichThumbnail` with image providers
- New `ImmichLocalThumbnailProvider` for showing local thumbnails
- Loads low resolution thumbnail in multipart loading (need to test the usefulness of this)
- Adds blur effect to thumbnail previews in gallery mode
- Changes custom image provider type to be their own class
- Adds blurhash (Closes #7055 )

### Thumbnail preview 

Without blur / With blur
 

Also might add in cache for remote images in this one.

### CODE REVIEW

**Comment by martyfuhry** [comment]
That blur is a bit aggressive, maybe it could be taken down a bit? It's difficult to find the "right value" which reduces the sharpness of the pixelated version without making it look blurry.

File: mobile/lib/shared/ui/fade_in_placeholder_image.dart [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return SizedBox.expand(
+      child: Stack(
+        fit: StackFit.expand,
+        children: [
+          placeholder,
+          FadeInImage(
+            fadeInDuration: const Duration(milliseconds: 100),
```
Comment (shenlong-tanwen) [inline mobile/lib/shared/ui/fade_in_placeholder_image.dart:26]:
```suggestion
            fadeInDuration: duration,
```

File: mobile/lib/shared/ui/hooks/blurhash_hook.dart [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ValueNotifier<Uint8List?> useBlurHashState(Asset? asset) {
+  if (asset?.thumbhash == null) {
+    return useState(null);
+  }
+
+  final rbga = thumbhash.thumbHashToRGBA(
+    base64Decode(asset!.thumbhash!),
+  );
+
+  return useState(thumbhash.rgbaToBmp(rbga));
```
Comment (shenlong-tanwen) [inline mobile/lib/shared/ui/hooks/blurhash_hook.dart:17]: Why do we need this to be a state? Can we use `useRef` instead?

**Review by shenlong-tanwen — APPROVED** [review]
LGTM! Love the ImageProvider related PRs <3

File: mobile/lib/shared/ui/hooks/blurhash_hook.dart (hunk 2 above)
Comment (martyfuhry) [reply mobile/lib/shared/ui/hooks/blurhash_hook.dart]: Didn't know the difference! Thanks for the suggestion. I've made the change, and I think it'll be very helpful.

File: mobile/lib/shared/ui/fade_in_placeholder_image.dart (hunk 1 above)
Comment (martyfuhry) [reply mobile/lib/shared/ui/fade_in_placeholder_image.dart]: Thank you, I missed this!

**Comment by alextran1502** [comment]
Hi @martyfuhry, I just tested this PR, and here are a few thoughts.

- When dragging across the timeline, I found the gradient placeholder is more prominent than the blur hash image. I wonder if the gradient placeholder should be replaced entirely by the blurhash.
- If the point above cannot be solved, I found the transition from the gradient -> blurhash -> actual image is very fast. I don't see the need to add blurhash and increase the complexity.

I think my point is to try to keep the complexity down if we can when the trade-off is minimal. Let me know what you think

**Comment by fyfrey** [comment]
Hi @alextran1502 

we can't show the blurhash while drag scrolling because we don't load the asset information. We only fetch this information from the database once drag scrolling stops.

The transition is indeed very fast for cached thumbnails. For non-cached images it makes a lot of sense, though. We can probably improve this in the future to only use the blurhash if the image is not cashed

**Comment by martyfuhry** [comment]
Google photos doesn't show a blurhash at all on my device. I agree it is very very fast and may appear too busy. We could replace the placeholders entirely with blurhashes if we can query the assets while scrubbing quickly enough to read their blurhash.

**Pull request merged**